src/cache_invalidate.c

```diff
@@ -59,45 +60,44 @@ cache_invalidate_relcache_all(void)
 	ts_bgw_job_cache_invalidate_callback();
 }
 
+static Oid hypertable_proxy_table_oid = InvalidOid;
+static Oid bgw_proxy_table_oid = InvalidOid;
+
+void
+ts_cache_invalidate_set_proxy_tables(Oid hypertable_proxy_oid, Oid bgw_proxy_oid)
+{
+	hypertable_proxy_table_oid = hypertable_proxy_oid;
+	bgw_proxy_table_oid = bgw_proxy_oid;
+}
+
 /*
  * This function is called when any relcache is invalidated.
  * Should route the invalidation to the correct cache.
+ *
+ * NOTE that the callback should not call any functions that could invoke the
+ * relcache or syscache to query information during the invalidation. That
+ * might lead to bad things happening.
  */
 static void
-cache_invalidate_callback(Datum arg, Oid relid)
+cache_invalidate_relcache_callback(Datum arg, Oid relid)
 {
-	static bool in_recursion = false;
-
-	if (ts_extension_invalidate(relid))
+	if (relid == InvalidOid)
```

Comment:
```suggestion
	if (!OidIsValid(relid))
```
Doesn't this branch get kind of expensive when taken? For all those callbacks, even if no timescaledb instance has been created you create and destroy 1 memory context per callback.

Reply:
This was already there in the old code. AFAIK, a "wildcard' invalidation is rarely issued, but needs to be handled nevertheless. This is the only place in the PG code where I could find such an invalidation:

```c
/*
 * CacheInvalidateRelcacheAll
 *		Register invalidation of the whole relcache at the end of command.
 *
 * This is used by alter publication as changes in publications may affect
 * large number of tables.
 */
void
CacheInvalidateRelcacheAll(void)
{
	PrepareInvalidationState();

	RegisterRelcacheInvalidation(InvalidOid, InvalidOid);
}

```